File: road_maintenance.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// 合并重叠的区间
vector<pair<int, int>> mergeIntervals(vector<pair<int, int>>& intervals) {
    if (intervals.empty()) {
        return {};
    }
    
    // 按开始时间排序
    sort(intervals.begin(), intervals.end());
    
    vector<pair<int, int>> merged;
    merged.push_back(intervals[0]);
    
    for (size_t i = 1; i < intervals.size(); i++) {
        // 如果当前区间与最后一个合并区间重叠，则合并
        if (intervals[i].first <= merged.back().second + 1) { // +1 因为时间点是连续的整数
            merged.back().second = max(merged.back().second, intervals[i].second);
        } else {
            merged.push_back(intervals[i]);
        }
    }
    
    return merged;
}

// 计算可用的维修时间
long long calculateAvailableTime(int H, vector<pair<int, int>>& warningPeriods) {
    if (warningPeriods.empty()) {
        return H;
    }
    
    // 合并重叠的预警时段
    vector<pair<int, int>> mergedWarnings = mergeIntervals(warningPeriods);
    
    // 计算被预警时段覆盖的总时间
    long long coveredTime = 0;
    for (const auto& period : mergedWarnings) {
        coveredTime += period.second - period.first + 1;
    }
    
    // 可用时间 = 总时间 - 被覆盖时间
    long long availableTime = H - coveredTime;
    
    return max(0LL, availableTime);
}

// 解决道路维修时间计算问题
void solveRoadMaintenance() {
    int T; // 测试用例数量
    cin >> T;
    
    for (int t = 0; t < T; t++) {
        int H; // 总工作时间
        int N; // 预警时段数量
        cin >> H >> N;
        
        vector<pair<int, int>> warningPeriods;
        for (int i = 0; i < N; i++) {
            int xi, yi;
            cin >> xi >> yi;
            warningPeriods.push_back({xi, yi});
        }
        
        // 计算并输出结果
        long long result = calculateAvailableTime(H, warningPeriods);
        cout << result << endl;
    }
}

// 测试函数
void testExample() {
    cout << "测试示例:" << endl;
    cout << "输入:" << endl;
    cout << "1" << endl;
    cout << "10" << endl;
    cout << "2" << endl;
    cout << "5 10" << endl;
    cout << "5 5" << endl;
    cout << endl;
    cout << "预期输出: 4" << endl;
    cout << endl;
    
    // 模拟输入
    int H = 10;
    vector<pair<int, int>> warningPeriods = {{5, 10}, {5, 5}};
    
    long long result = calculateAvailableTime(H, warningPeriods);
    cout << "实际输出: " << result << endl;
    cout << "测试" << (result == 4 ? "通过" : "失败") << endl;
}

int main() {
    // 运行测试
    testExample();
    cout << "\n" << string(50, '=') << endl;
    cout << "运行主程序:" << endl;
    cout << "请输入测试用例数据:" << endl;
    
    // 运行主程序
    solveRoadMaintenance();
    
    return 0;
}

File: road_maintenance.py
def solve_road_maintenance():
    """
    解决道路维修时间计算问题
    输入：T个测试用例，每个包含H（总时间）和N个预警时段
    输出：每个测试用例的可用维修时间
    """
    
    def merge_intervals(intervals):
        """
        合并重叠的区间
        输入：intervals - 区间列表，每个区间为 [start, end]
        输出：合并后的区间列表
        """
        if not intervals:
            return []
        
        # 按开始时间排序
        intervals.sort(key=lambda x: x[0])
        
        merged = [intervals[0]]
        
        for current in intervals[1:]:
            # 如果当前区间与最后一个合并区间重叠，则合并
            if current[0] <= merged[-1][1] + 1:  # +1 因为时间点是连续的整数
                merged[-1][1] = max(merged[-1][1], current[1])
            else:
                merged.append(current)
        
        return merged
    
    def calculate_available_time(H, warning_periods):
        """
        计算可用的维修时间
        输入：H - 总时间，warning_periods - 预警时段列表
        输出：可用维修时间总数
        """
        if not warning_periods:
            return H
        
        # 合并重叠的预警时段
        merged_warnings = merge_intervals(warning_periods)
        
        # 计算被预警时段覆盖的总时间
        covered_time = 0
        for start, end in merged_warnings:
            covered_time += end - start + 1
        
        # 可用时间 = 总时间 - 被覆盖时间
        available_time = H - covered_time
        
        return max(0, available_time)
    
    # 读取输入
    T = int(input())  # 测试用例数量
    
    for _ in range(T):
        H = int(input())  # 总工作时间
        N = int(input())  # 预警时段数量
        
        warning_periods = []
        for _ in range(N):
            xi, yi = map(int, input().split())
            warning_periods.append([xi, yi])
        
        # 计算并输出结果
        result = calculate_available_time(H, warning_periods)
        print(result)


# 测试函数
def test_example():
    """
    测试示例用例
    """
    print("测试示例:")
    print("输入:")
    print("1")
    print("10")
    print("2")
    print("5 10")
    print("5 5")
    print()
    print("预期输出: 4")
    print()
    
    # 模拟输入
    H = 10
    warning_periods = [[5, 10], [5, 5]]
    
    def merge_intervals(intervals):
        if not intervals:
            return []
        intervals.sort(key=lambda x: x[0])
        merged = [intervals[0]]
        for current in intervals[1:]:
            if current[0] <= merged[-1][1] + 1:
                merged[-1][1] = max(merged[-1][1], current[1])
            else:
                merged.append(current)
        return merged
    
    def calculate_available_time(H, warning_periods):
        if not warning_periods:
            return H
        merged_warnings = merge_intervals(warning_periods)
        covered_time = 0
        for start, end in merged_warnings:
            covered_time += end - start + 1
        available_time = H - covered_time
        return max(0, available_time)
    
    result = calculate_available_time(H, warning_periods)
    print(f"实际输出: {result}")
    print(f"测试{'通过' if result == 4 else '失败'}")


if __name__ == "__main__":
    # 运行测试
    test_example()
    print("\n" + "="*50)
    print("运行主程序:")
    print("请输入测试用例数据:")
    solve_road_maintenance()
